### PR TITLE
Clustering

### PR DESCRIPTION
**Pull request by TonyBagnall**
something odd happened with the other clustering PR. Creating a new one

### CODE REVIEW

File: sktime/clustering/partitioning/_averaging_metrics.py
Comment (ABostrom) [inline sktime/clustering/partitioning/_averaging_metrics.py:169]: I would consider if you can optimise this using numpy vectorisation. 

Multiple nested loops will be slow, and i think you can definitely optimise this easily.

Comment (TonyBagnall) [reply sktime/clustering/partitioning/_averaging_metrics.py:169]: good point, can I suggest that once the PR is in, this could be raised as an issue?

File: sktime/clustering/_k_means.py
Comment (mloning) [inline sktime/clustering/_k_means.py:64]:
```suggestion
    _averaging_algorithm_dict: AveragingAlgoDict = {
```

File: sktime/clustering/_k_means.py
Comment (mloning) [inline sktime/clustering/_k_means.py:92]: We may want to consider in the future to add `y` which is always ignored to be able to reuse composition functionality from classification (e.g. pipelines or tuning). 
```suggestion
    def fit(self, X: NumpyOrDF, y : NumpyArray = None) -> None:
```

File: sktime/clustering/base/base.py
Comment (mloning) [inline sktime/clustering/base/base.py:52]: This isn't strictly necessary here as it's set in the constructor of the `BaseEstimator`
```suggestion
```

File: sktime/clustering/evaluation/_plot_clustering.py
Comment (mloning) [inline sktime/clustering/evaluation/_plot_clustering.py:16]: Looks like we need these imports here

File: sktime/clustering/partitioning/_averaging_metrics.py
Comment (mloning) [inline sktime/clustering/partitioning/_averaging_metrics.py:139]: For later improvements, we could either try to vectorise these loops or use numba.

File: sktime/clustering/partitioning/_center_initializers.py
Comment (mloning) [inline sktime/clustering/partitioning/_center_initializers.py:48]:
```suggestion
    def _initialize_centers(self) -> NumpyArray:
```

File: sktime/clustering/_k_means.py
Comment (mloning) [inline sktime/clustering/_k_means.py:127]:
```suggestion
    def _calculate_new_centers(self, cluster_values: NumpyArray) -> NumpyArray:
```

File: sktime/clustering/partitioning/_center_initializers.py
Comment (mloning) [inline sktime/clustering/partitioning/_center_initializers.py:115]: temp could probably be a numpy array here which we preallocate and then write into

File: sktime/clustering/partitioning/_cluster_approximations.py
Comment (mloning) [inline sktime/clustering/partitioning/_cluster_approximations.py:29]:
```suggestion
    def _approximate(self) -> int:
```

File: sktime/clustering/partitioning/_lloyds_partitioning.py
Comment (mloning) [inline sktime/clustering/partitioning/_lloyds_partitioning.py:207]:
```suggestion
    def _cluster_data(self, X: NumpyArray) -> List[List[int]]:
```

File: sktime/clustering/partitioning/_lloyds_partitioning.py
Comment (mloning) [inline sktime/clustering/partitioning/_lloyds_partitioning.py:222]:
```suggestion
    def _update_centers(self, data: NumpyArray):
```

File: sktime/tests/_config.py
Comment (mloning) [inline sktime/tests/_config.py:88]: Could we make `BaseCluster` available in the parent directory by importing it in the init file?  
```suggestion
from sktime.clustering.base import BaseCluster
```
Perhaps also rename to `BaseClusterer`?

File: sktime/utils/_testing/panel.py
Comment (mloning) [inline sktime/utils/_testing/panel.py:112]: Why is `X` a list here? Shouldn't we use `_make_panel_X(**kwargs)` here?

File: sktime/utils/clustering_utils.py
Comment (mloning) [inline sktime/utils/clustering_utils.py:1]: If they're only used for clustering I'd consider moving them into `sktime/clustering/base`

File: sktime/utils/_testing/panel.py
Comment (mloning) [reply sktime/utils/_testing/panel.py:112]: The signature also seems wrong, isn't it: `random.randint(low, high=None, size=None, ...)`?

**Review by mloning — COMMENTED** [review]
Thanks again @TonyBagnall @chrisholder 

Just a few points, happy to merge after fixing them:
* please avoid double underscores
* one strange line in `sktime/utils/_testing/panel.py` (see comments below)

The other points are all minor, happy to ignore them for now.

I didn't look at the example notebooks and files in the contrib folder. 

Going forward, we should think about setting up a testing framework after merging this PR.

File: sktime/clustering/_k_means.py
Comment (chrisholder) [reply sktime/clustering/_k_means.py:64]: This must have got readded by mistake when i was fixing whatever happened with the pr the other day good spot

File: sktime/clustering/base/base.py
Comment (chrisholder) [reply sktime/clustering/base/base.py:52]: This must have got readded by mistake when i was fixing whatever happened with the pr the other day good spot

File: sktime/clustering/partitioning/_center_initializers.py
Comment (chrisholder) [reply sktime/clustering/partitioning/_center_initializers.py:48]: This must have got readded by mistake when i was fixing whatever happened with the pr the other day good spot

File: sktime/clustering/partitioning/_cluster_approximations.py
Comment (chrisholder) [reply sktime/clustering/partitioning/_cluster_approximations.py:29]: This must have got readded by mistake when i was fixing whatever happened with the pr the other day good spot

File: sktime/clustering/partitioning/_lloyds_partitioning.py
Comment (chrisholder) [reply sktime/clustering/partitioning/_lloyds_partitioning.py:222]: This must have got readded by mistake when i was fixing whatever happened with the pr the other day good spot

File: sktime/clustering/partitioning/_lloyds_partitioning.py
Comment (chrisholder) [reply sktime/clustering/partitioning/_lloyds_partitioning.py:207]: This must have got readded by mistake when i was fixing whatever happened with the pr the other day good spot

File: sktime/clustering/partitioning/_lloyds_partitioning.py
Comment (TonyBagnall) [reply sktime/clustering/partitioning/_lloyds_partitioning.py:222]: and its spelt centres ....

File: sktime/clustering/partitioning/_center_initializers.py
Comment (chrisholder) [reply sktime/clustering/partitioning/_center_initializers.py:48]: This one actually needs to be public because I want to be able to call this from outside the class. So you construct the center initlizers and then when you want new centers you call initialize_centers()

File: sktime/clustering/partitioning/_cluster_approximations.py
Comment (chrisholder) [reply sktime/clustering/partitioning/_cluster_approximations.py:29]: Same here want this to be a publically avaliable method

File: sktime/clustering/_k_means.py
Comment (chrisholder) [reply sktime/clustering/_k_means.py:127]: Same here this wants to be public

File: sktime/clustering/partitioning/_lloyds_partitioning.py
Comment (chrisholder) [reply sktime/clustering/partitioning/_lloyds_partitioning.py:222]: Right big design decision here. I've gone with the American spelling (and I did double check since you got me questioning my whole existance with that spelling https://www.grammarly.com/blog/center-centre/). If we rename this we should also go change CenterInitializer and any other American spellings.

Comment (TonyBagnall) [reply sktime/clustering/partitioning/_lloyds_partitioning.py:222]: Lol, leave as is for now, I can sneakily refactor another time. Seriously though, American spelling is probably better, as much as it galls me to say it

**Review by mloning — APPROVED** [review]
Hi @chrisholder @TonyBagnall - so I think this is good to go, I reviewed twice and the most important points have been addressed - thanks @chrisholder! Happy to merge this unless someone else wants to take another look?

**Comment by TonyBagnall** [comment]
> Hi @chrisholder @TonyBagnall - so I think this is good to go, I reviewed twice and the most important points have been addressed - thanks @chrisholder! Happy to merge this unless someone else wants to take another look?

please get it in, thanks.

File: sktime/clustering/base/base.py
Comment (fkiraly) [inline sktime/clustering/base/base.py:126]: why does this need a mixin? Just put it in the base class.

**Pull request merged**
